# 01. Importing Libraries

In [1]:
# Import libraries

import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

# 02. Importing Data

In [2]:
# Import the data set into Jupyter as a new dataframe
path = r'/Users/ChuahLH/Desktop/Documents/Instacart Basket Analysis'

In [3]:
# Import the orders_products_merged dataframe
customers = pd.read_csv(os.path.join(path, '02 Data','Original Data', 'customers.csv'))

In [4]:
# Investigate dataframe
customers.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [5]:
customers.shape

(206209, 10)

# 03. Rename Columns

In [7]:
df_customers = customers

In [8]:
df_customers.rename(columns = {'First Name':'first_name'}, inplace = True)

In [9]:
df_customers.rename(columns = {'Surnam':'last_name'}, inplace = True)

In [10]:
df_customers.rename(columns = {'Gender':'gender'}, inplace = True)

In [11]:
df_customers.rename(columns = {'STATE':'state'}, inplace = True)

In [12]:
df_customers.rename(columns = {'Age':'age'}, inplace = True)

In [13]:
df_customers.rename(columns = {'n_dependants':'number_of_dependants'}, inplace = True)

In [14]:
df_customers.rename(columns = {'fam_status':'family_status'}, inplace = True)

In [15]:
df_customers.head()

user_id first_name last_name  gender       state  age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   number_of_dependants family_status  income  
0                     3       married  165665  
1                     0        single   59285  
2                     2       married   99568  
3                     0        single   42049  
4                     1       married   40374

In [16]:
df_customers.shape

(206209, 10)

# 04. Consistency Checks

In [18]:
# Check for duplications
df_customers[df_customers.duplicated()]

Empty DataFrame
Columns: [user_id, first_name, last_name, gender, state, age, date_joined, number_of_dependants, family_status, income]
Index: []

It is found no duplication in data frame. 

In [21]:
# Check for missing data
df_customers.isnull().sum()

user_id                     0
first_name              11259
last_name                   0
gender                      0
state                       0
age                         0
date_joined                 0
number_of_dependants        0
family_status               0
income                      0
dtype: int64

There are 11259 missing data - first_name.

In [23]:
# View missing values in first_name column
df_nan_first_name = df_customers[df_customers['first_name'].isnull() == True]

In [24]:
df_nan_first_name

user_id first_name last_name  gender         state  age date_joined  \
53        76659        NaN   Gilbert    Male      Colorado   26    1/1/2017   
73        13738        NaN     Frost  Female     Louisiana   39    1/1/2017   
82        89996        NaN    Dawson  Female        Oregon   52    1/1/2017   
99        96166        NaN   Oconnor    Male      Oklahoma   51    1/1/2017   
105       29778        NaN    Dawson  Female          Utah   63    1/1/2017   
...         ...        ...       ...     ...           ...  ...         ...   
206038   121317        NaN    Melton    Male  Pennsylvania   28   3/31/2020   
206044   200799        NaN  Copeland  Female        Hawaii   52    4/1/2020   
206090   167394        NaN     Frost  Female        Hawaii   61    4/1/2020   
206162   187532        NaN     Floyd  Female    California   39    4/1/2020   
206171   116898        NaN   Delgado  Female      Colorado   23    4/1/2020   

        number_of_dependants family_status  income  
53                         2       married   41709  
73                         0        single   82518  
82                         3       married  117099  
99                         1       married  155673  
105                        3       married  151819  
...                      ...           ...     ...  
206038                     3       married   87783  
206044                     2       married  108488  
206090                     1       married   45275  
206162                     0        single   56325  
206171                     2       married   59222  

[11259 rows x 10 columns]

It is approx. 11259 / 206209 x 100% = 5% data missing. Leave the data within the data frame. 

In [37]:
# Check for the descriptive statistics - count, mean, standard deviation, minimum, lower quartile, median, upper quartile and maximum
df_customers.describe()


user_id            age  number_of_dependants         income
count  206209.000000  206209.000000         206209.000000  206209.000000
mean   103105.000000      49.501646              1.499823   94632.852548
std     59527.555167      18.480962              1.118433   42473.786988
min         1.000000      18.000000              0.000000   25903.000000
25%     51553.000000      33.000000              0.000000   59874.000000
50%    103105.000000      49.000000              1.000000   93547.000000
75%    154657.000000      66.000000              3.000000  124244.000000
max    206209.000000      81.000000              3.000000  593901.000000

In [38]:
# Check for data types of the columns
df_customers.dtypes

user_id                  int64
first_name              object
last_name               object
gender                  object
state                   object
age                      int64
date_joined             object
number_of_dependants     int64
family_status           object
income                   int64
dtype: object

In [40]:
# Checking for mixed data types 
for col in df_customers.columns.tolist():
  weird = (df_customers[[col]].applymap(type) != df_customers[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_customers[weird]) > 0:
    print (col)

No Mixed Data Types. 

# 05. Combine Data

In [42]:
# Import the orders_products_merged dataframe
orders_products_merged = pd.read_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_merged_updated_rev_02.pkl'))

In [44]:
# Check the orders_products_merged output

orders_products_merged.head(5)

Unnamed: 0_x  order_id  user_id  order_number  orders_day_of_week  \
0             0   2539329        1             1                   2   
1             1   2398795        1             2                   3   
2             2    473747        1             3                   3   
3             3   2254736        1             4                   4   
4             4    431534        1             5                   4   

   order_hour_of_day  days_since_last_order  product_id  add_to_cart_order  \
0                  8                    NaN       196.0                1.0   
1                  7                   15.0       196.0                1.0   
2                 12                   21.0       196.0                1.0   
3                  7                   29.0       196.0                1.0   
4                 15                   28.0       196.0                1.0   

   reordered  ...     busiest_day         busiest_days  busiest_period_of_day  \
0        0.0  ...  Regularly busy  Regularly busy days         Average orders   
1        1.0  ...      Least busy      Least busy days         Average orders   
2        1.0  ...      Least busy      Least busy days            Most orders   
3        1.0  ...      Least busy      Least busy days         Average orders   
4        1.0  ...      Least busy      Least busy days            Most orders   

   max_order  loyalty_flag mean_order spending_flag  \
0         10  New customer   6.367797   Low spender   
1         10  New customer   6.367797   Low spender   
2         10  New customer   6.367797   Low spender   
3         10  New customer   6.367797   Low spender   
4         10  New customer   6.367797   Low spender   

  median_days_since_last_order          frequent_flag         Frequency_flag  
0                    20.259259  Non-frequent customer  Non-Frequent customer  
1                    20.259259  Non-frequent customer  Non-Frequent customer  
2                    20.259259  Non-frequent customer  Non-Frequent customer  
3                    20.259259  Non-frequent customer  Non-Frequent customer  
4                    20.259259  Non-frequent customer  Non-Frequent customer  

[5 rows x 27 columns]

In [46]:
orders_products_merged.shape

(32406041, 27)

Review both data frame. Shared key column is user_id. 

In [47]:
# Check for data types of the columns
orders_products_merged.dtypes

Unnamed: 0_x                       int64
order_id                           int64
user_id                            int64
order_number                       int64
orders_day_of_week                 int64
order_hour_of_day                  int64
days_since_last_order            float64
product_id                       float64
add_to_cart_order                float64
reordered                        float64
Unnamed: 0_y                       int64
product_name                      object
aisle_id                           int64
department_id                      int64
prices                           float64
_merge                          category
price_range_loc                   object
busiest_day                       object
busiest_days                      object
busiest_period_of_day             object
max_order                          int64
loyalty_flag                      object
mean_order                       float64
spending_flag                     object
median_days_sinc

Both data types are same for user_id. 

In [48]:
# Merge the dataframe

df_merged = orders_products_merged.merge(df_customers, on = 'user_id')

In [49]:
# Investigate the dataframe
df_merged.head(5)

Unnamed: 0_x  order_id  user_id  order_number  orders_day_of_week  \
0             0   2539329        1             1                   2   
1             1   2398795        1             2                   3   
2             2    473747        1             3                   3   
3             3   2254736        1             4                   4   
4             4    431534        1             5                   4   

   order_hour_of_day  days_since_last_order  product_id  add_to_cart_order  \
0                  8                    NaN       196.0                1.0   
1                  7                   15.0       196.0                1.0   
2                 12                   21.0       196.0                1.0   
3                  7                   29.0       196.0                1.0   
4                 15                   28.0       196.0                1.0   

   reordered  ...         Frequency_flag first_name  last_name  gender  \
0        0.0  ...  Non-Frequent customer      Linda     Nguyen  Female   
1        1.0  ...  Non-Frequent customer      Linda     Nguyen  Female   
2        1.0  ...  Non-Frequent customer      Linda     Nguyen  Female   
3        1.0  ...  Non-Frequent customer      Linda     Nguyen  Female   
4        1.0  ...  Non-Frequent customer      Linda     Nguyen  Female   

     state age date_joined number_of_dependants family_status income  
0  Alabama  31   2/17/2019                    3       married  40423  
1  Alabama  31   2/17/2019                    3       married  40423  
2  Alabama  31   2/17/2019                    3       married  40423  
3  Alabama  31   2/17/2019                    3       married  40423  
4  Alabama  31   2/17/2019                    3       married  40423  

[5 rows x 36 columns]

In [50]:
df_merged.shape

(32406041, 36)

# 06. Export Data Frame as Pickle File

In [51]:
# Export new file as pkl
df_merged.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'df_merged_final.pkl'))